### ISSUE TITLE
All Item Processor Wiki Sections Lacks XML Tutorial

### ISSUE THREAD

**Issue by TurtleShroom** [Vanilla-Expanded/VanillaExpandedFramework#68]
Good Afternoon

I wanted to add a new Combination Def, but the Wiki only demonstrated the C-Sharp and pointed me to existing Combinations. The problem is that none of the examples gave me what I needed to know; I wanted an item to require TWO inputs, but nothing shows me how to do that.

Please consider eludicating, like you did with the Animal Behaviors section, the full list of potential XML statements.

Thank you.
-TURTLESHROOM

**Comment by legodude17** [comment]
The C# shows all the fields and what they do, no?

**Comment by TurtleShroom** [comment]
> The C# shows all the fields and what they do, no?

I can't read C-Sharp. I tried doing it, but I can't figure how what's going on.

Here's an example:

![image](https://github.com/Vanilla-Expanded/VanillaExpandedFramework/assets/17663470/6dc0543e-7766-4c72-805e-8134bada616e)

(Please ignore the Blue Light Filter.)

I tried to change the Soda Fountain to accept four different inputs, with an option to haev only one filled, two filled, and so on.

HJowever, the Wiki did not explain to me how to do that.

`
					<li Class="ItemProcessor.CompProperties_ItemProcessor">
						<buildingOnGraphic>Things/Buildings/VBE_SodaDispenser</buildingOnGraphic>
						<numberOfInputs>4</numberOfInputs>
						<acceptsNoneAsInput>true</acceptsNoneAsInput>
						<transfersIngredientLists>true</transfersIngredientLists>
						<isSemiAutomaticMachine>true</isSemiAutomaticMachine>
						<noPowerDestroysProgress>false</noPowerDestroysProgress>
						<destroyIngredientsAtStartOfProcess>false</destroyIngredientsAtStartOfProcess>
						<destroyIngredientsAtAwfulQuality>false</destroyIngredientsAtAwfulQuality>
						<showProgressBar>true</showProgressBar>
						<mustLoadFromInteractionSpot>false</mustLoadFromInteractionSpot>
						<mustUnloadFromInteractionSpot>false</mustUnloadFromInteractionSpot>
						<isAutoDropper>true</isAutoDropper>
						<rareTicksToDestroy>20</rareTicksToDestroy>
						
						<resetSemiautomaticIcon>UI/VBE_SyrupReset</resetSemiautomaticIcon>
						<chooseIngredientsIcon>UI/VBE_EmptySyrupIcon</chooseIngredientsIcon>
						<bringIngredientsText>VBE_StartInsertion</bringIngredientsText>
						<bringIngredientsDesc>VBE_StartInsertionDesc</bringIngredientsDesc>
						<bringIngredientsIcon>UI/VBE_InsertSyrup</bringIngredientsIcon >
						<cancelIngredientsText>VBE_CancelBringingSyrup</cancelIngredientsText> 
						<cancelIngredientsDesc>VBE_CancelBringingSyrupDesc</cancelIngredientsDesc>
						<cancelIngredientsIcon>UI/VBE_CancelSyrup</cancelIngredientsIcon> 
						
						<InsertFirstItemDesc>VBE_InsertSyrupDesc</InsertFirstItemDesc>
						<InsertSecondItemDesc>VBE_InsertSyrupDesc</InsertSecondItemDesc>
						<InsertThirdItemDesc>VBE_InsertSyrupDesc</InsertThirdItemDesc>
						<InsertFourthItemDesc>VBE_InsertSyrupDesc</InsertFourthItemDesc>
						
					</li>
`

There is no documentation on this.

Also, when I made multiple Item Accepted Defs, if they shared the same ingredient in the same slot, I got this error:

``
Exception filling window for Verse.FloatMenu: System.NullReferenceException: Object reference not set to an instance of an object
  at ItemProcessor.Building_ItemProcessor.<IngredientsChosenBringThemIn>b__83_3 (ItemProcessor.CombinationDef element) [0x0002b] in <b3d96b3c2a5f43a5a7194d98b33781f3>:0 
  at System.Linq.Enumerable+WhereListIterator`1[TSource].MoveNext () [0x00037] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.TryGetFirst[TSource] (System.Collections.Generic.IEnumerable`1[T] source, System.Boolean& found) [0x00045] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at System.Linq.Enumerable.First[TSource] (System.Collections.Generic.IEnumerable`1[T] source) [0x00000] in <351e49e2a5bf4fd6beabb458ce2255f3>:0 
  at ItemProcessor.Building_ItemProcessor.IngredientsChosenBringThemIn () [0x00889] in <b3d96b3c2a5f43a5a7194d98b33781f3>:0 
  at ItemProcessor.Command_SetSecondItemList.TryInsertSecondThing (System.Boolean category, System.String item) [0x00149] in <b3d96b3c2a5f43a5a7194d98b33781f3>:0 
  at ItemProcessor.Command_SetSecondItemList+<>c__DisplayClass4_0.<ProcessInput>b__4 () [0x0001c] in <b3d96b3c2a5f43a5a7194d98b33781f3>:0 
  at Verse.FloatMenuOption.Chosen (System.Boolean colonistOrdering, Verse.FloatMenu floatMenu) [0x00030] in <95de19971c5d40878d8742747904cdcd>:0 
  at Verse.FloatMenuOption.DoGUI (UnityEngine.Rect rect, System.Boolean colonistOrdering, Verse.FloatMenu floatMenu) [0x00535] in <95de19971c5d40878d8742747904cdcd>:0 
  at Verse.FloatMenu.DoWindowContents (UnityEngine.Rect rect) [0x000fb] in <95de19971c5d40878d8742747904cdcd>:0 
  at Verse.Window.InnerWindowOnGUI (System.Int32 x) [0x001d3] in <95de19971c5d40878d8742747904cdcd>:0 
UnityEngine.StackTraceUtility:ExtractStackTrace ()
(wrapper dynamic-method) Verse.Log:Verse.Log.Error_Patch1 (string)
Verse.Window:InnerWindowOnGUI (int)
UnityEngine.GUI:CallWindowDelegate (UnityEngine.GUI/WindowFunction,int,int,UnityEngine.GUISkin,int,single,single,UnityEngine.GUIStyle)

**Comment by TurtleShroom** [comment]
This is the Patch I tried to write, but it doesn't work and I can't troubleshoot it because the Wiki lacks documentation for ALL of the XML lines.

[TSP_VFEB_SodaIncorporation.zip](https://github.com/Vanilla-Expanded/VanillaExpandedFramework/files/12826201/TSP_VFEB_SodaIncorporation.zip)

**Comment by TurtleShroom** [comment]
What frustrates my efforts the most is that _**NOTHING**_ explains how the "_<acceptsNoneAsInput>true</acceptsNoneAsInput>_" Boolean qualifier works, and the Wiki is silent on it.

**Comment by TurtleShroom** [comment]
![image](https://github.com/Vanilla-Expanded/VanillaExpandedFramework/assets/17663470/44c907eb-4348-49d1-84be-7a112eb630d1)

**Comment by databobek** [comment]
you have to search for acceptNoneAsInput, instead of a XML node. It says that the given ingredient is allowed to be empty.

**Comment by TurtleShroom** [comment]
I tried that, but Sarg said it did not work.
--
Sent from Mail.ru app for Android Sunday, 15 October 2023, 08:07AM -04:00 from databobek  ***@***.*** :

**Comment by TurtleShroom** [comment]
> you have to search for acceptNoneAsInput, instead of a XML node. It says that the given ingredient is allowed to be empty.

I don't understand. Sarg said that the C-Sharp Boolean statement didn't work when I asked him about this. Is there some other way to do this?

**Comment by TurtleShroom** [comment]
> you have to search for acceptNoneAsInput, instead of a XML node. It says that the given ingredient is allowed to be empty.

How do I do this? What does that mean?

**Comment by databobek** [comment]
> > you have to search for acceptNoneAsInput, instead of a XML node. It says that the given ingredient is allowed to be empty.
> 
> How do I do this? What does that mean?

I am not saying it works, it's described in the code as "allowing an input to be empty"

When looking for a node from xml <NODENAME> you don't look in the code for <NODENAME> but only NODENAME.

**Comment by TurtleShroom** [comment]
Right, but when I try that, it does not work. That is a serious flaw in the VFE system thar really needs fixing.
--
Sent from Mail.ru app for Android Wednesday, 25 October 2023, 03:30AM -04:00 from databobek  ***@***.*** :

**Comment by legodude17** [comment]
The whole Item Processor system needs a rewrite, but no one's had the time or will to do that yet. Given VEF is developed for VE mods, it's a bit weird to call this a "serious flaw" that "really needs fixing" when no VE mod uses the feature.

**Comment by TurtleShroom** [comment]
I strongly disagree. Countless non-VFE Mods use this Mod's system, and you would think not forcing people to make separate buildings to process one ingredient versus a whole other, separate instance for two inputs, would be common sense.
--
Sent from Mail.ru app for Android Wednesday, 25 October 2023, 05:33PM -04:00 from Joshua  ***@***.*** :

**Comment by legodude17** [comment]
I'm not saying it's not an issue, I'm just saying you're being rather hyperbolic. It's one small feature of an outdated system, not a massive flaw in the entire framework.
